lib/Clam/Properties/BndCheck.cc

```diff
@@ -0,0 +1,348 @@
+    llvm::Value *op;
+    unsigned sz;
+    if (isa<AllocaInst>(I)) {
+      AllocaInst *allocaI = dyn_cast<AllocaInst>(&I);
+      op = allocaI->getArraySize();
+      const llvm::DataLayout *dl = getInsDatalayout(*allocaI);
+      sz = dl->getTypeSizeInBits(allocaI->getAllocatedType()) / 8;
+    } else {
+      op = I.getOperand(0);
+    }
+    llvm::Value &operand = *op;
+    crab_lit_ref_t opd = m_lfac.getLit(operand);
+    if (!opd->isInt()) {
+      CLAM_WARNING("Unexpected type of alloc size operand");
+      return;
+    }
+    assert(opd->isInt());
+    auto Iopd = std::static_pointer_cast<const crabIntLit>(opd);
+    unsigned width = Iopd->getBitwidth();
+    var_t ref_obj_sz = m_lfac.mkIntVar(width);
+    if (isa<AllocaInst>(I))
+      bb.assign(ref_obj_sz, m_lfac.getExp(opd) * sz);
+    else
+      bb.assign(ref_obj_sz, m_lfac.getExp(opd));
+    // create a var_t for offset and assign it to zero
+    var_t ref_offset = m_lfac.mkIntVar(width);
+    bb.assign(ref_offset, number_t(0));
+
+    // add pair to the map by llvm pointer
+    m_ref_bnd_map.insert({&I, {ref_obj_sz, ref_offset}});
+  }
+
+  // -- get element ptr additional utilities construction
+  // From LLVM IR to crab IR we perform the following transformation:
+  // p := gep(ptr, offset) => p := gep_ref(region, ptr + offset)
+  // We use the idea from fat pointer like by adding two additional variables:
+  // p_offset := ptr_offset + offset;
+  // p_size := ptr_size;
+  // where p_offset indicates the offset of element pointer
+  // the p_size means the total size of allcation
+  // the element ptr should refere the same size as the original one
+  void visitAfterGep(llvm::Instruction &I, CrabGepRefOps &s) {
+    if (!isa<PHINode>(I) && !isa<BitCastInst>(I) &&
+        !isa<GetElementPtrInst>(I)) {
+      return;
+    }
+    lin_exp_t offset = s.getOffset();
+    basic_block_t &bb = s.getBasicBlock();
+    CRAB_LOG("cfg-bnd-check",
+             llvm::errs() << "Gep: add two extra variable for " << I << "\n");
+    Value *vlhs = llvm::cast<llvm::Value>(&I);
+    Value *vptr = I.getOperand(0);
+    auto it = m_ref_bnd_map.find(vptr);
+    if (it == m_ref_bnd_map.end()) {
+      CLAM_ERROR("Could not find gep ptr "
+                 << *vptr << " from reference map \n Stop at" << I);
+    } else {
+      var_t ptr_obj_sz = it->second.first;
+      var_t ptr_offset = it->second.second;
+      var_t lhs_obj_sz =
+          m_lfac.mkIntVar(ptr_obj_sz.get_type().get_integer_bitwidth());
+      var_t lhs_offset =
+          m_lfac.mkIntVar(ptr_offset.get_type().get_integer_bitwidth());
+      bb.assign(lhs_obj_sz, ptr_obj_sz);
+      bb.assign(lhs_offset, offset);
+      m_ref_bnd_map.insert({&I, {lhs_obj_sz, lhs_offset}});
+    }
+  }
+};
+
+EmitBndChecks::EmitBndChecks(const CrabBuilderParams &params,
+                             crabLitFactory &lfac, uint32_t &assertionId) {
+  this->ref_bnd_map = std::make_unique<ref_bnd_map_t>();
```

Comment:
`ref_bnd_map` should go to `EmitBndChecksImpl`. 
The idea of using the pimp patterns is to completely hide all the class fields and methods.